pyquil/pyqvm.py

```diff
@@ -0,0 +1,459 @@
+
+        elif isinstance(instruction, Measurement):
+            if not run_and_measure:
+                measured_val = self.wf_simulator.do_measurement(qubit=instruction.qubit.index)
+                x = instruction.classical_reg  # type: MemoryReference
+                self.ram[x.name][x.offset] = measured_val
+            else:
+                # Hacky code to speed up run-and-measure programs
+                # Don't actually do the measurement, just make a note of where in ram the bits
+                # will go
+                x = instruction.classical_reg  # type: MemoryReference
+                self._ram_measure_mapping[instruction.qubit.index] = (x.name, x.offset)
+
+                if self.program_counter == self._last_measure_program_loc:
+                    # We've reached the last measure instruction. Time to sample from our
+                    # wavefunction
+                    bitstrings = self.wf_simulator.sample_bitstrings(self.program.num_shots)
+
+                    # Quil2 doesn't support defining multidimensional arrays, and its typical
+                    # to allocate a readout register of size n_bits and assume "the stack" will
+                    # do the right thing and give you an array of shape (n_shots, n_bits) instead.
+                    # Here we resize all of our readout registers to be this extended 2d array shape
+                    ro_registers = sorted(set(ram_name for ram_name, ram_offset
+                                              in self._ram_measure_mapping.values()))
+                    for ro_register in ro_registers:
+                        assert np.sum(self.ram[ro_register]) == 0, 'reading out into a parameter?'
+                        prev_shape = self.ram[ro_register].shape
+                        assert len(prev_shape) == 1, prev_shape
+                        prev_shape = prev_shape[0]
+                        prev_dtype = self.ram[ro_register].dtype
+                        assert prev_dtype == QUIL_TO_NUMPY_DTYPE['BIT'], prev_dtype
+                        self.ram[ro_register] = np.zeros((self.program.num_shots, prev_shape),
+                                                         dtype=prev_dtype)
+
+                    # Penultimately, we use our collected qubit-to-ram mappings to fill in our newly
+                    # reshaped ram arrays
+                    for q in range(bitstrings.shape[1]):
+                        if q in self._ram_measure_mapping:
+                            ram_name, ram_offset = self._ram_measure_mapping[q]
+                            self.ram[ram_name][:, ram_offset] = bitstrings[:, q]
+
+                    # Finally, we RESET the system because it isn't mandated yet that programs
+                    # contain RESET instructions.
+                    self.wf_simulator.reset()
+            self.program_counter += 1
+
+        elif isinstance(instruction, Declare):
+            self.ram[instruction.name] = np.zeros(instruction.memory_size,
```

Comment:
How does the offset get stored? Should we add a disclaimer that `SHARING` is not supported?